lib/Dialect/SV/Transforms/HWRaiseInOutPorts.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+void HWInOutPortConversion::buildInputSignals() {
+  if (hasReaders()) {
+    Value readValue =
+        converter.createNewInput(origPort, "_rd", origPort.type, readPort);
+
+    // Replace all sv::ReadInOutOp's with the new input.
+    Value origInput = body->getArgument(origPort.argNum);
+    for (auto *user : llvm::make_early_inc_range(origInput.getUsers())) {
+      sv::ReadInOutOp read = dyn_cast<sv::ReadInOutOp>(user);
+      if (!read)
+        continue;
+
+      read.replaceAllUsesWith(readValue);
+      read.erase();
+    }
+  }
+
+  if (hasWriters()) {
+    sv::AssignOp write = writers.front();
+    converter.createNewOutput(origPort, "_wr", origPort.type, write.getSrc(),
+                              writePort);
+    write.erase();
+  }
+}
+
+void HWInOutPortConversion::buildOutputSignals() {
+  // TODO: could support hw.inout outputs (always create read/write ports) -
+  // don't need it for now, though.
+  assert(false && "hw.inout outputs not yet supported");
+}
+
+void HWInOutPortConversion::mapInputSignals(OpBuilder &b, Operation *inst,
+                                            Value instValue,
+                                            SmallVectorImpl<Value> &newOperands,
+                                            ArrayRef<Backedge> newResults) {
+
+  if (hasReaders()) {
+    // Create a read_inout op at the instantiation point. This effectively
+    // pushes the read_inout op from the module to the instantiation site.
+    newOperands[readPort.argNum] =
+        b.create<ReadInOutOp>(inst->getLoc(), instValue).getResult();
+  }
+
+  if (hasWriters()) {
+    // Create a sv::AssignOp at the instantiation point. This effectively
+    // pushes the write op from the module to the instantiation site.
+    Value writeFromInsideMod = newResults[writePort.argNum];
+    b.create<sv::AssignOp>(inst->getLoc(), instValue, writeFromInsideMod);
+  }
+}
+
+void HWInOutPortConversion::mapOutputSignals(
+    OpBuilder &b, Operation *inst, Value instValue,
+    SmallVectorImpl<Value> &newOperands, ArrayRef<Backedge> newResults) {
+  llvm_unreachable("hw.inout outputs not yet supported");
+}
+
+class HWInoutPortConversionBuilder : public PortConversionBuilder {
+public:
+  using PortConversionBuilder::PortConversionBuilder;
+  FailureOr<std::unique_ptr<PortConversion>> build(hw::PortInfo port) override {
+    if (port.direction == hw::PortDirection::INOUT)
+      return {std::make_unique<HWInOutPortConversion>(converter, port)};
+    return PortConversionBuilder::build(port);
+  }
+};
+
+} // namespace
+
+void HWRaiseInOutPortsPass::runOnOperation() {
+  // Find all modules and run port conversion on them.
+  circt::hw::InstanceGraph &instanceGraph =
+      getAnalysis<circt::hw::InstanceGraph>();
+  llvm::DenseSet<InstanceGraphNode *> visited;
+  auto res = instanceGraph.getInferredTopLevelNodes();
+
+  // Visit the instance hierarchy in a depth-first manner, modifying child
```

Comment:
Not for this PR, but this feels like something which could be pulled out. Perhaps a part of the `PortConversion` framework... IIRC the ESI pass does something like this as well.

Reply:
that, or some kind of boring utility library (global/hierarchy modification) which can be coupled together with PortConvesion (local changes).